### ISSUE TITLE
Utility that prints out delayed values when they're computed?

### ISSUE THREAD

**Issue by gjoseph92** [dask/dask#9859]
It's common to see `print(dask_thing.compute())` in the middle of people's code. Print statements are helpful for debugging, and especially with dask, where everything is delayed already, it can be especially helpful to see what an object actually looks like. However, these `compute` calls can enormously slow things down and recompute data unnecessarily, and users may not understand this.

What if we had a utility for making print statements like this that basically deferred the printing/evaluation until a later call of `dask.compute`?

#### Original code:
```python
Q3 = ddf["trip_time"].quantile(0.75)
print(f"Q3 is:  {Q3.compute()}")  # blocks, computes whole dataframe, prints
upper_bound = Q3 + (1.5 * (Q3 - lower_bound))
print(f"Upper bound is:  {upper_bound.compute()}")  # blocks, recomputes whole dataframe, prints

ddf = ddf.loc[(ddf["trip_time"] >= lower_bound) & (ddf["trip_time"] <= upper_bound)]
ddf.to_parquet(...).compute()  # blocks, does actual computation you care about
```

#### Using new utility:
```python
Q3 = ddf["trip_time"].quantile(0.75)
dask.print("Q3 is:  {}", Q3)  # instant, doesn't print anything yet
upper_bound = Q3 + (1.5 * (Q3 - lower_bound))
dask.print("Upper bound is:  {}", upper_bound)  # instant, doesn't print anything yet

ddf = ddf.loc[(ddf["trip_time"] >= lower_bound) & (ddf["trip_time"] <= upper_bound)]
ddf.to_parquet(...).compute()  # blocks, does actual computation you care about.
# first prints "Q3 is ..."
# then a little later prints "Upper bound is ..."
# then later returns
```

Basically, `dask.print` / `dask.log` (not sure what to name it) would just push a print statement onto an internal, global queue when you call it, but not actually print anything yet.

Then each time it's called, `dask.compute` would pull everything off that queue and add it to the list of dask collections to be computed. It would take Futures to each of the print statements, and as soon as the future completed, it would print.

**Comment by jsignell** [comment]
I really like this idea. It gets at an issue that I have struggled with before: https://github.com/dask/dask/discussions/7202 there is also the similar sounding https://distributed.dask.org/en/stable/api.html#distributed.print which is different, but we should know how to explain that difference.

**Comment by gjoseph92** [comment]
> There is also the similar sounding https://distributed.dask.org/en/stable/api.html#distributed.print which is different, but we should know how to explain that difference.

Yeah, not sure about naming (that's why I didn't pick one yet). Maybe `dask.log`, `dask.record`, `dask.debug`?

**Comment by fjetter** [comment]
A think a simple way to achieve this is to just wrap the calling function with something like

```python
def _remote_print_results(func, format_string="", *args, **kwargs):
    from distributed import print
    res = func(*args, **kwargs)
    if format_string:
        print(format_string)
    else:
        print(res)
    return res
```

see https://distributed.dask.org/en/stable/api.html#distributed.print

We should ensure that the function name of a wrapped function is the same as before to not throw off the dashboard, etc.

**Comment by hendrikmakait** [comment]
> Re: name. I don't think we'll want to call this `print`. Maybe `log` as suggested above or even something verbose as `log_result`, etc.

FWIW, I would rather not use `log*` for something unrelated to actual logging but more of a delayed `distributed.print` callback.

**Comment by j-bennet** [comment]
POC here: https://github.com/dask/dask/pull/10321

**Comment by fjetter** [comment]
I thought about this problem again and my earlier suggestion was quite nonsense. @gjoseph92 already had a proper suggestion about the "future on a stack and compute pops this off, etc.". Implementing this is rather straight forward and looks like this

```diff
diff --git a/dask/base.py b/dask/base.py
index c9e0491b7..ac5b9b8a6 100644
--- a/dask/base.py
+++ b/dask/base.py
@@ -575,10 +575,13 @@ def compute(
     >>> dask.compute({'a': a, 'b': b, 'c': 1})
     ({'a': 45, 'b': 4.5, 'c': 1},)
     """
-
     collections, repack = unpack_collections(*args, traverse=traverse)
     if not collections:
         return args
+    global _PRINT_QUEUE
+    if _PRINT_QUEUE:
+        collections.extend(_PRINT_QUEUE)
+        _PRINT_QUEUE = []

     schedule = get_scheduler(
         scheduler=scheduler,
@@ -1532,3 +1535,21 @@ def clone_key(key, seed):
         prefix = key_split(key)
         return prefix + "-" + tokenize(key, seed)
     raise TypeError(f"Expected str or tuple[str, Hashable, ...]; got {key}")
+
+
+_PRINT_QUEUE = []
+
+
+def print_on_compute(msg="%s", val=None):
+    from dask import delayed
+
+    @delayed
+    def _print_future(res):
+        try:
+            from distributed.worker import print
+        except ImportError:
+            pass
+        print(msg % res)
+
+    _PRINT_QUEUE.append(_print_future(val))
+    return val
```

In this example implementation I haven't looked into any kind of edge cases and I expect there to be a few (e.g. the print futures should likely only be popped off if we're actually computing the result).

This looks quite promising. It would still impact graph optimization and such but for _content_ debugging this would still be very powerful.

@j-bennet care to pick this up again and hammer out the details?

**Comment by j-bennet** [comment]
@fjetter Sure, I'll take it.
